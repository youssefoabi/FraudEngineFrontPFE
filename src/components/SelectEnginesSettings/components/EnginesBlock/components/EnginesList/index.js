import { fromPairs, pipe, prop, sortBy, values } from 'ramda';
import { func, shape } from 'prop-types';
import { isNilOrEmpty } from 'ramda-adjunct';
import React, { useMemo } from 'react';

import DraggableList from '../../../../../DragableList';
import EngineChip from './components/EngineChip';

const EnginesList = ({ enginesGroup, handleChange }) => {
  const { engines } = enginesGroup || {};
  const items = useMemo(() => pipe(values, sortBy(prop('order')))(engines), [engines]);

  const reorderEngine = (engine, index) => {
    if (isNilOrEmpty(engine)) return null;

    const { engineId } = engine;
    return [engineId, { ...(engines[engineId] || {}), order: index }];
  };

  const onOrderChange = newEnginesOrder => {
    if (isNilOrEmpty(newEnginesOrder)) return;

    const newEngines = newEnginesOrder.map(reorderEngine);
    handleChange({ id: 'engines', value: fromPairs(newEngines) });
  };

  const renderEngine = item => {
    if (isNilOrEmpty(item)) return null;

    const { item: currentEngine } = item;
    if (isNilOrEmpty(currentEngine)) return null;

    return (
      <EngineChip
        engineId={prop('engineId', currentEngine)}
        enginesGroup={enginesGroup}
        handleChange={handleChange}
      />
    );
  };

  if (isNilOrEmpty(items)) return null;

  return (
    <DraggableList
      droppableId="engines"
      onOrderChange={onOrderChange}
      data={items}
      renderItem={renderEngine}
      keyExtractor={prop('engineId')}
    />
  );
};

EnginesList.propTypes = {
  enginesGroup: shape({}),
  handleChange: func,
};

EnginesList.defaultProps = {
  enginesGroup: null,
  handleChange: Function.Prototype,
};

export default EnginesList;
