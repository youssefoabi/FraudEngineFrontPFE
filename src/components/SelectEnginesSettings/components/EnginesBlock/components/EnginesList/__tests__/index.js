import { itRendersAllMutations } from '@cdiscount/jest-utils';

import EnginesList from '../';

const mutations = [
  {
    name: 'with engines',
    props: {
      enginesGroup: {
        engines: {
          1: {
            engineId: 1,
            engineName: 'engine 1',
            order: 0,
            isEnable: true,
            engineLabel: 'engineLabel',
            pictogramUrl: 'pictogramUrl',
          },
          2: {
            engineId: 2,
            engineName: 'engine 2',
            order: 1,
            isEnable: true,
            engineLabel: 'engineLabel 2',
            pictogramUrl: 'pictogramUrl 2',
          },
        },
      },
    },
  },
  { name: 'with null engines' },
];

describe(EnginesList.name, () => {
  itRendersAllMutations(EnginesList, mutations);
});
