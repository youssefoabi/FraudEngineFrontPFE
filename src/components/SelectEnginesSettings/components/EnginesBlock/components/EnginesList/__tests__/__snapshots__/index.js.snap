// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnginesList EnginesList renders correctly with engines 1`] = `
<DraggableList
  data={
    Array [
      Object {
        "engineId": 1,
        "engineLabel": "engineLabel",
        "engineName": "engine 1",
        "isEnable": true,
        "order": 0,
        "pictogramUrl": "pictogramUrl",
      },
      Object {
        "engineId": 2,
        "engineLabel": "engineLabel 2",
        "engineName": "engine 2",
        "isEnable": true,
        "order": 1,
        "pictogramUrl": "pictogramUrl 2",
      },
    ]
  }
  droppableId="engines"
  keyExtractor={[Function]}
  onOrderChange={[Function]}
  renderItem={[Function]}
/>
`;

exports[`EnginesList EnginesList renders correctly with null engines 1`] = `""`;
